### ISSUE TITLE
Inserts with column lists fail in Spark

### ISSUE THREAD

**Issue by alondhe** [OHDSI/WebAPI#1433]
_A little context -- I am working on getting the OHDSI stack up and running at Amgen. As we use a Spark based platform, I've developed a branch of SqlRender that supports Spark SQL. Thus far, I've gotten everything in Atlas working except Cohort Characterization_

As discussed in the FE repo (https://github.com/OHDSI/FeatureExtraction/issues/82), inserts with column lists are not supported in Spark SQL. In my SqlRender branch, I have to remove the column lists from these commands. However, this means that now the data to insert must match the expected columns in the table exactly. This poses issues with the Cohort Characterization service in Atlas, because there are many inserts into cc_results that rely upon column lists, with the payload being certain columns rather than all columns.

I'm wondering if we could come to consensus on altering the way we write insert commands in WebAPI (and OHDSI at large) to help with this issue, and to simply have more precise SQL. 

**My proposal is that all data to be inserted must have fields that correspond with the target table. It would ensure compatibility for all dialects and is more explicit.**

Thoughts? @anthonysena @chrisknoll @pavgra 

Also tagging @wivern  -- I've heard Hive has the same limitation.

**Comment by alondhe** [comment]
To clarify, the column lists can stay, as I am able to remove them in SqlRender. But the payload must have all fields present.

**Comment by anthonysena** [comment]
@alondhe from our offline discussion, I understand the practical need you have around these INSERT commands. The only downside I see is that by explicitly providing all values for all columns in the insert, we lose the ability to use default values defined at the table level. That said, looking at the results schema DDL, we utilize defaults in a limited manner and it sounds like that would be incompatible with Spark anyways. We could simply parameterize the SQL and use defaults provided in the code that renders the SQL so I suppose that is easy enough to handle.

**Comment by chrisknoll** [comment]
I'm sorry to be a nay-sayer on this, but inserts with column names is an added level of protection you have to changes in underlying DDL changes. For example, let's say one version of a DDL has the order modified_date and created_date, and due to some cleanup/code reorganizing, a dev thinks it would be better to define the DDL as create_date, modified date.

This would be fine:
```
insert into some_table (col1, col2, col3, created_date, modified_date) 
select col1, '1' as col2, col3, item_created_date, item_modified_date)  from some_temp_table
```

this would reverse the column values:
```
insert into some_table
select col1, '1' as col2, col3, item_created_date, item_modified_date)  from some_temp_table
```

In addition, if you add new, optional columns to a result schema, queries would have to be modified to explicitly set `null`.

I think this is something that the SparkSQL team should just get on top of and support missing columns/ordered columns in insert statement.  I don't think we should compromise the integrity of our queries to support some DBMS vendor shortcoming.

Just my 2 cents, but I will have a pretty hard stance on this with respect to circe queries, but of course the leadership behind WebAPI and FeatureExtraction can decide how they wish to compromise their libraries ;)

**Comment by alondhe** [comment]
Makes sense @chrisknoll. A workaround I'm brainstorming is to instead intercept the query execution and then obtain the table metadata (using Spark's stored proc of `describe table`), and then piece together the fields as appropriate.

Not sure yet how this will work.

**Comment by chrisknoll** [comment]
Yeah, I hear ya and can imagine the underlying difficulties of requiring the soruce select to match the destination table, but I think you're touching on where the solution lies: when the Spark engine gets the request, in part of parsing it it should convert that statement out to the proper order of columns and fill in any missing columns with null, and then let it through to the underlying engine.  It is in this 'parsing and ordering' phase of the query where it could notify the caller of invalid columns, etc, which is my main point above about 'integrity'.  

Hopefully the Spark community can pull it together to make it work. It does look like a cool system.

**Comment by alondhe** [comment]
Thanks to @anthonysena and @chrisknoll for the helpful discussion on this topic. And @chrisknoll -- appreciate you digging into this on Stack Overflow :) 

I've engaged with our vendors, but nothing imminent in terms of fixing this at the Spark or JDBC level. In the interim, I've modified our WebAPI's CancelableJdbcTasklet class to use a modified SqlRender to intercept inserts and re-work them to align with the table structure.
